c7n/resources/route53.py

```diff
@@ -194,6 +194,35 @@ def process_resource_set(self, client, domains, keys):
                 TagsToDelete=keys)
 
 
+@HostedZone.action_registry.register('delete')
+class Delete(BaseAction):
+    """Action to delete Route 53 hosted zones
+
+    It is recommended to use a filter to avoid unwanted deletion of R53 hosted zones
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: route53-delete-testing-hosted-zones
+                resource: aws.hostedzone
+                filters:
+                  - 'tag:TestTag': present
+                actions:
+                  - type: delete
+    """
+
+    schema = type_schema('delete')
+    permissions = ('route53:DeleteHostedZone',)
+
+    def process(self, hosted_zones):
+        client = local_session(self.manager.session_factory).client('route53')
+        for hz in hosted_zones:
+            self.manager.retry(
+                client.delete_hosted_zone, Id=hz['Id'])
```

Comment:
we should pass in `ignore_err_codes=('NoSuchHostedZone', 'NoSuchHostedZone')` to retry so that we handle idempotent exceptions. one question, glancing over the exception codes, is it accurate that we can't delete a zone if it has a record in it?

Reply:
Thanks for the suggestion. Added it to the code.
With regards to:

> one question, glancing over the exception codes, is it accurate that we can't delete a zone if it has a record in it?

Yes, that is true. If you try to delete a hosted zone that contains records that are not the NS / SOA ones you'll get:

```
botocore.errorfactory.HostedZoneNotEmpty: An error occurred (HostedZoneNotEmpty) when calling the DeleteHostedZone operation: The specified hosted zone contains non-required resource record sets  and so cannot be deleted.
2020-09-07 17:52:41,592: custodian.commands:ERROR The following policies had errors while executing . . .
```
So I guess I will have to ignore the error code `HostedZoneNotEmpty` and add support for r53domain deletion, so I can empty the Route53 zone before deleting it

I'll need to craft the

Reply:
hm, it seems I got confused by the r53domains, they're a totally different thing and they are not related with the hosted zones.

It is interesting, we should have some mechanism to empty a hostedzone before it's cleaned-up, and I think this should be part of Custodian. If not, the delete operation won't be completed.

Reply:
so I guess the "HostedZone" delete operation should also be in charge of deleting all of the record sets that are inside of the HostedZone prior deleting it.

Reply:
Just added the clean-up of the resource record sets before the clean-up of the hosted zone is done.

I _think_ this is done :)